### PR TITLE
Rewrite Http2 test to be less flaky

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
Existing Failure 

```
HttpOverHttp2Test > responseHeadersAfterGoaway(Protocol, MockWebServer) > okhttp3.internal.http2.HttpOverHttp2Test.responseHeadersAfterGoaway(Protocol, MockWebServer)[1] FAILED
    org.opentest4j.AssertionFailedError: 
    expected: "ABC"
     but was: null
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at okhttp3.internal.http2.HttpOverHttp2Test.responseHeadersAfterGoaway(HttpOverHttp2Test.java:1814)
```

Ongoing saga of https://github.com/square/okhttp/issues/4836

**Comment by yschimke** [comment]
New failure

```
HttpOverHttp2Test > responseHeadersAfterGoaway(Protocol, MockWebServer) > okhttp3.internal.http2.HttpOverHttp2Test.responseHeadersAfterGoaway(Protocol, MockWebServer)[2] FAILED
    okhttp3.internal.http2.StreamResetException: stream was reset: PROTOCOL_ERROR
        at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.kt:148)
        at okhttp3.internal.http2.Http2ExchangeCodec.readResponseHeaders(Http2ExchangeCodec.kt:97)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:110)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
```

**Comment by yschimke** [comment]
@swankjesse landing as it seems to reduce flakes after some builds.  Will observe, and follow up with any concerns or review comments.

File: okhttp/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java [hunk 1]
```diff
@@ -1789,30 +1793,46 @@ public void connectionShutdownAfterHealthCheck(
...
     client.newCall(new Request.Builder().url(server.url("/")).build()).enqueue(callback);
 
-    assertThat(bodies.poll(3, SECONDS)).isEqualTo("DEF");
-    assertThat(bodies.poll(3, SECONDS)).isEqualTo("ABC");
-    assertThat(server.getRequestCount()).isEqualTo(2);
+    latch.await();
+
+    assertThat(bodies.remove()).isEqualTo("DEF");
+
+    if (errors.isEmpty()) {
```
Comment (amirlivneh) [inline okhttp/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java:1825]: I don't know the root cause for the flakiness in this case, but I suggest that you look at a few techniques I used in the past to make similar tests deterministic and that may apply here: https://github.com/square/okhttp/pull/4792, https://github.com/square/okhttp/pull/4728, https://github.com/square/okhttp/pull/4772.

Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java:1825]: Thanks, will do. I wanted to get something in so we have PR and master with working builds.  So I hope you forgive my hackish attempt to start!

Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java:1825]: I will get familiar with QueueDispatcher, there are a couple more flaky tests here.

**Review by swankjesse — COMMENTED** [review]
LGTM

I prefer LinkedBlockingDeque over CountDownLatch for tests like this; I think it’s easier to follow.

**Comment by yschimke** [comment]
I'm due to come back and fix this, it's sitting in my github notification queue taunting me for months now.  But glad I've fixed for now, even in a hacky way, as my priority was green PR builds.

I will get to it though :)

**Pull request merged**